internal/resources/product/model.go

```diff
@@ -0,0 +1,470 @@
+			Attributes: pie.Map(productVariant.Attributes, NewProductVariantAttribute),
+			Prices:     pie.Map(productVariant.Prices, NewProductPriceDraft),
+			Staged:     utils.BoolRef(false),
+		})
+	}
+
+	// changeMasterVariant
+	if !p.MasterVariant.ID.Equal(plan.MasterVariant.ID) {
+		result.Actions = append(result.Actions, platform.ProductChangeMasterVariantAction{
+			Sku:    plan.MasterVariant.Sku.ValueStringPointer(),
+			Staged: utils.BoolRef(false),
+		})
+	}
+
+	// removeVariant
+	for _, productVariant := range variantsRemoved {
+		result.Actions = append(result.Actions, platform.ProductRemoveVariantAction{
+			Sku:    productVariant.Sku.ValueStringPointer(),
+			Staged: utils.BoolRef(false),
+		})
+	}
+
+	// # Compare Product Variant attributes and prices
+	for _, currentVariant := range currentVariants {
+		sku := currentVariant.Sku.ValueString()
+		planVariantRef := getProductVariantBySku(planVariants, sku)
+		if planVariantRef != nil {
+			// Process Attributes
+			attributesAdded, attributesModified, attributesRemoved := compare(currentVariant.Attributes, planVariantRef.Attributes, "Name")
+			for _, attribute := range append(attributesAdded, attributesModified...) {
+				// setAttribute
+				result.Actions = append(result.Actions, platform.ProductSetAttributeAction{
+					Sku:    &sku,
+					Name:   attribute.Name.ValueString(),
+					Staged: utils.BoolRef(false),
+					Value:  unmarshalAttributeValue(attribute.Value.ValueString()),
+				})
+			}
+			for _, attribute := range attributesRemoved {
+				// removeAttribute
+				result.Actions = append(result.Actions, platform.ProductSetAttributeAction{
+					Sku:    &sku,
+					Name:   attribute.Name.ValueString(),
+					Staged: utils.BoolRef(false),
+				})
+			}
+
+			// #Process Prices
+			pricesAdded, pricesModified, pricesRemoved := compare(currentVariant.Prices, planVariantRef.Prices, "Key")
+
+			// addPrice
+			for _, price := range pricesAdded {
+				result.Actions = append(result.Actions, platform.ProductAddPriceAction{
+					Sku:    &sku,
+					Price:  NewProductPriceDraft(price),
+					Staged: utils.BoolRef(false),
+				})
+			}
+
+			// changePrice
+			for _, price := range pricesModified {
+				result.Actions = append(result.Actions, platform.ProductChangePriceAction{
+					PriceId: price.ID.ValueString(),
+					Price:   NewProductPriceDraft(price),
+					Staged:  utils.BoolRef(false),
+				})
+			}
+
+			// removePrice
+			for _, price := range pricesRemoved {
+				result.Actions = append(result.Actions, platform.ProductRemovePriceAction{
+					Sku:     &sku,
+					PriceId: price.ID.ValueString(),
+					Staged:  utils.BoolRef(false),
+				})
+			}
+		}
+	}
+
+	return result
+}
+
+func unmarshalAttributeValue(value string) any {
+	var data any
+	json.Unmarshal([]byte(value), &data)
+	return data
+}
+
+func marshalAttributeValue(o platform.Attribute) string {
+	val, err := json.Marshal(o.Value)
+	if err != nil {
+		panic(err)
```

Comment:
Can we return this error instead of panic-ing? It allows terraform to deal with the failing resource a bit more elegantly (this might stop the whole application midway processing, which might cause side-effects when other processes are running in parallel)

Reply:
I will fix this.